File: javascript/retry.js
function servicioPenca() {
    print("Servicio invocado");
    if (Math.random() < 0.8) {
	throw "GuateoException";
    }
    print("Servicio ejecutado exitosamente");
}

function solucion() {
    retry({times: 10, when: "GuateoException", wait: 2}, servicioPenca);
}

function retry(options, callback) {
    var lastException = null;
    var times = options.times || 10;
    var exception = options.when;
    var wait = options.wait;
    while (times-- > 0) {
        try {
            return callback();
	} catch (e) {
	    if (!exception || exception == e) {
		sleep(wait);
		lastException = e;
	    } else {
		throw e;
	    }
	}
    }
    throw lastException;
}

// No sleep funcion in JS so we implement it with a busy loop
function sleep(seconds) {
    var start = new Date().getTime();
    while(true) {
	if (new Date().getTime() - start > (seconds * 1000)) break;
    }
}

solucion();

File: python/retry.py
import time, random

class GuateoException(Exception): pass

def servicio_penca():
  print "Servicio invocado"
  if random.random() < 0.8:
      raise GuateoException()
  print "Servicio ejecutado exitosamente"


def solucion():
    retry(servicio_penca, when=GuateoException, times=10, wait=2)

def retry(callable, when, times, wait):
    last_exception = None
    while times > 0:
        times -= 1
        try:
            return callable()
        except when, exception:
            time.sleep(wait)
            last_exception = exception
    raise last_exception

if __name__ == '__main__':
    solucion()
